stan/math/prim/scal/err/is_size_match.hpp

```diff
@@ -0,0 +1,30 @@
+#ifndef STAN_MATH_PRIM_SCAL_ERR_IS_SIZE_MATCH_HPP
+#define STAN_MATH_PRIM_SCAL_ERR_IS_SIZE_MATCH_HPP
+
+#include <boost/type_traits/common_type.hpp>
+#include <stan/math/prim/scal/meta/likely.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Check if the provided sizes match
+ *
+ * @tparam T_size1 Type of size 1
+ * @tparam T_size2 Type of size 2
+ *
+ * @param i Size 1
+ * @param j Size 2
+ *
+ * @return <code>true</code> if provided dimensions match
+ */
+template <typename T_size1, typename T_size2>
+inline bool is_size_match(T_size1 i, T_size2 j) {
```

Comment:
Question... do you know if `check_size_match()` is actually used this way? I'm wondering why we don't just pass in the variables... there's probably a good reason for it.

Reply:
No, I based `is_size_match()` around passing the replicated test cases from `check_size_match()`.

Reply:
thanks. I was able to find its uses. It's actually used when we check things like the number of columns of a variable matches the number of rows of another.

Reply:
Ah sorry I thought you were asking why the sizes are passed directly into `likely(i == static_cast<T_size1>(j))` rather than by some other method.

Reply:
I looked into this closer. Clearly this is optimized in ways that some other `check` functions may not be. Many are passing entire structures and then calculating simple `.size()` operations. 

Would it be desirable to change the use of all the `check` functions to pass the calculated values (i.e `check_fun(y.size())` rather than `check_fun(y)`) where possible?

Reply:
I'm a little behind on the rationale for this design pattern. Would the uses of `check_size_match` be changed to `check_matching_size`? Since `check_matching_size` simply passes the size values of the referenced structure into `check_size_match` anyway?

Reply:
Yes, the code is beyond me in those cases, but I now see what you are saying.